[BUG][2.1.2b1] Fail to restore sub-path when repository is quoted

## Bug summary

rdiff-backup beta version cannot restore special filename "Fichier @ <root>".

## Version, Python, Operating System

python 3.10
rdiff-backup 2.1.2b1
Linux Debian

## rdiff-backup call

**With 2.1.2b1**
```
$ rdiff-backup -v 5 --restore-as-of=1454448640 "/home/ikus060/Downloads/testcases/testcases/Fichier @ <root>" /home/ikus060/Downloads/coucou

WARNING: this command line interface is deprecated and will disappear, start using the new one as described with '--new --help'.
* Unable to import module (py)xattr. Extended attributes not supported on filesystem at path /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
* Unable to import module posix1e from pylibacl package. POSIX ACLs not supported on filesystem at path /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
* Unable to import win32security module. Windows ACLs not supported by filesystem at path /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
* Could not determine if source directory /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data permits trailing spaces or periods in filenames because we can't find any files with trailing dot/period. It will be treated as permitting such files, but none will exist if it doesn't, so it doesn't really matter and is harmless
* -----------------------------------------------------------------
Detected abilities for rdiff-backup repository (read only) file system:
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
* Unable to import module (py)xattr. Extended attributes not supported on filesystem at path /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
* Unable to import module posix1e from pylibacl package. POSIX ACLs not supported on filesystem at path /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
* Unable to import win32security module. Windows ACLs not supported by filesystem at path /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
* -----------------------------------------------------------------
Detected abilities for restore target (read/write) file system:
  Ownership changing                           Off
  Hard linking                                 On
  fsync() directories                          On
  Directory inc permissions                    On
  High-bit permissions                         On
  Symlink permissions                          Off
  Extended filenames                           On
  Windows reserved filenames                   Off
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
WARNING: The special escapes file '/home/ikus060/Downloads/testcases/testcases/rdiff-backup-data/special_escapes' was not found, will assume need to escape DOS devices and trailing spaces based on file systems
* Backup: escape_dos_devices = 0
* Backup: escape_trailing_spaces = 0
* Cleaning up
Traceback (most recent call last):
  File "/home/ikus060/workspace/PDSL/rdiffweb.git/.venv/bin/rdiff-backup", line 32, in <module>
    sys.exit(rdiffbackup.run.main_run(sys.argv[1:]))
  File "/home/ikus060/workspace/PDSL/rdiffweb.git/.venv/lib/python3.10/site-packages/rdiffbackup/run.py", line 86, in main_run
    ret_val |= conn_act.setup()
  File "/home/ikus060/workspace/PDSL/rdiffweb.git/.venv/lib/python3.10/site-packages/rdiffbackup/actions/restore.py", line 150, in setup
    self.repo.setup_quoting()
  File "/home/ikus060/workspace/PDSL/rdiffweb.git/.venv/lib/python3.10/site-packages/rdiffbackup/locations/repository.py", line 217, in setup_quoting
    self.ref_path = FilenameMapping.get_quotedrpath(self.ref_path)
  File "/home/ikus060/workspace/PDSL/rdiffweb.git/.venv/lib/python3.10/site-packages/rdiff_backup/FilenameMapping.py", line 180, in get_quotedrpath
    assert not rp.index, (
AssertionError: Trying to start quoting '/home/ikus060/Downloads/testcases/testcases/Fichier @ <root>' in the middle.
```

**With 2.0.5**
```
$ rdiff-backup -v 5 --restore-as-of=1454448640 "/home/ikus060/Downloads/testcases/testcases/Fichier @ <root>" /home/ikus060/Downloads/coucou

Using rdiff-backup version 2.0.5
        with cpython /home/ikus060/workspace/PDSL/rdiffweb.git/.venv/bin/python version 3.10.4
        on Linux-5.17.15-76051715-generic-x86_64-with-glibc2.35, fs encoding utf-8
Unable to import module (py)xattr.
Extended attributes not supported on filesystem at /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
Unable to import module posix1e from pylibacl package.
POSIX ACLs not supported on filesystem at /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
Unable to import win32security module. Windows ACLs
not supported by filesystem at /home/ikus060/Downloads/testcases/testcases/rdiff-backup-data
-----------------------------------------------------------------
Detected abilities for rdiff-backup repository (read only) file system:
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
Unable to import module (py)xattr.
Extended attributes not supported on filesystem at /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
Unable to import module posix1e from pylibacl package.
POSIX ACLs not supported on filesystem at /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
Unable to import win32security module. Windows ACLs
not supported by filesystem at /home/ikus060/Downloads/coucou/rdiff-backup.tmp.0
-----------------------------------------------------------------
Detected abilities for restore target (read/write) file system:
  Ownership changing                           Off
  Hard linking                                 On
  fsync() directories                          On
  Directory inc permissions                    On
  High-bit permissions                         On
  Symlink permissions                          Off
  Extended filenames                           On
  Windows reserved filenames                   Off
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 Off
  Case sensitivity                             On
  Escape DOS devices                           Off
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
Warning: special_escapes file not found,
will assume need to escape DOS devices and trailing spaces based on file systems.
Backup: escape_dos_devices = 0
Backup: escape_trailing_spaces = 0
Starting restore of /home/ikus060/Downloads/testcases/testcases/Fichier @ <root> to /home/ikus060/Downloads/coucou as it was as of Tue Feb  2 16:30:40 2016.
Processing changed file .
Restore finished
```

## What happened and what did you expect?

Stack trace displayed when trying to restore.

## More information

I have similar problem with filename similar to `Char ;090 to quote`

My testcases repo could be downloaded from rdiffweb repo:
https://gitlab.com/ikus-soft/rdiffweb/-/blob/master/rdiffweb/tests/testcases.tar.gz

--- a/src/rdiff_backup/FilenameMapping.py
+++ b/src/rdiff_backup/FilenameMapping.py
@@ -177,13 +177,13 @@ def unquote(path):
 
 def get_quotedrpath(rp, separate_basename=0):
     """Return quoted version of rpath rp"""
-    assert not rp.index, (
-        "Trying to start quoting '{rp}' in the middle.".format(rp=rp))
     if separate_basename:
+        assert not rp.index, (
+            "Trying to start quoting '{rp}' in the middle.".format(rp=rp))
         dirname, basename = rp.dirsplit()
         return QuotedRPath(rp.conn, dirname, (unquote(basename), ), rp.data)
     else:
-        return QuotedRPath(rp.conn, rp.base, (), rp.data)
+        return QuotedRPath(rp.conn, rp.base, rp.index, rp.data)
 
 
 def _init_quoting_regexps():

--- a/src/rdiffbackup/locations/map/filenames.py
+++ b/src/rdiffbackup/locations/map/filenames.py
@@ -156,13 +156,13 @@ def get_quotedrpath(rp, separate_basename=0):
     """
     Return quoted version of rpath rp
     """
-    assert not rp.index, (
-        "Trying to start quoting '{rp}' in the middle.".format(rp=rp))
     if separate_basename:
+        assert not rp.index, (
+            "Trying to start quoting '{rp}' in the middle.".format(rp=rp))
         dirname, basename = rp.dirsplit()
         return QuotedRPath(rp.conn, dirname, (unquote(basename), ), rp.data)
     else:
-        return QuotedRPath(rp.conn, rp.base, (), rp.data)
+        return QuotedRPath(rp.conn, rp.base, rp.index, rp.data)
 
 
 def get_quoting_regexps(chars_to_quote, quoting_char):

--- a/testing/action_backuprestore_test.py
+++ b/testing/action_backuprestore_test.py
@@ -80,6 +80,27 @@ class ActionBackupRestoreTest(unittest.TestCase):
         # all tests were successful
         self.success = True
 
+    def test_action_backuprestore_quoted(self):
+        """
+        test the backup and restore actions with quoted repository
+        """
+        # we backup using a specific chars-to-quote
+        self.assertEqual(comtst.rdiff_backup_action(
+            False, False, self.from1_path, self.bak_path,
+            ("--api-version", "201", "--current-time", "10000",
+             "--chars-to-quote", "A"),
+            b"backup", ()), 0)
+
+        # then we restore once the full repo, once a sub-path
+        self.assertEqual(comtst.rdiff_backup_action(
+            True, False, os.path.join(self.bak_path, b"itemX"), self.to1_path,
+            ("--api-version", "201"),
+            b"restore", ()), 0)
+        self.assertEqual(comtst.rdiff_backup_action(
+            True, True, self.bak_path, self.to2_path,
+            ("--api-version", "201"),
+            b"restore", ()), 0)
+
     def tearDown(self):
         # we clean-up only if the test was successful
         if self.success:

